File: php/testcode/csrf_008.php
<?php
require_once __DIR__ . '/shared.php';

// vuln-code-snippet start php_csrf_predictable_token
function csrf008(BenchmarkRequest $req): BenchmarkResponse {
    $userId = $req->cookie('user_id');
    $token = md5($userId); // vuln-code-snippet vuln-line php_csrf_predictable_token
    if ($req->post('csrf_token') !== $token) {
        return BenchmarkResponse::badRequest('Invalid CSRF token');
    }
    return BenchmarkResponse::ok('Action performed');
}
// vuln-code-snippet end php_csrf_predictable_token

File: rust/testcode/csrf_008.rs
//! CWE-352: CSRF token only logged, not validated, before action is performed.

fn tracing_log(msg: &str) {
    let _ = msg;
}

fn perform_action(action: &str) -> bool {
    let _ = action;
    true
}

// vuln-code-snippet start testcodeCsrf008
pub fn handle(req: &super::shared::BenchmarkRequest) -> super::shared::BenchmarkResponse {
    tracing_log(&format!("csrf={}", req.param("csrf_token"))); // logged but not validated
    let action = req.param("action");
    let result = perform_action(&action); // vuln-code-snippet target-line testcodeCsrf008
    if result {
        super::shared::BenchmarkResponse::ok("action performed")
    } else {
        super::shared::BenchmarkResponse::error("action failed")
    }
}
// vuln-code-snippet end testcodeCsrf008
